core/winnt/src/TWinNTSystem.cxx

```diff
@@ -1914,44 +1913,52 @@ int  TWinNTSystem::MakeDirectory(const char *name)
 #endif
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Struct used to pass information between OpenDirectory and GetDirEntry in a
+/// thread safe way (each thread creates a new instance of it).
+
+struct FindFileData_t {
+   HANDLE          fSearchFile;         // HANDLE returned by FindFirstFile and used by FindNextFile
+   WIN32_FIND_DATA fFindFileData;       // Structure to look for files (aka OpenDir under UNIX)
+   Bool_t          fFirstFile{kFALSE};  // Flag used by OpenDirectory/GetDirEntry
+};
+
 ////////////////////////////////////////////////////////////////////////////////
 /// Close a WinNT file system directory.
 
 void TWinNTSystem::FreeDirectory(void *dirp)
 {
-   TSystem *helper = FindHelper(0, dirp);
-   if (helper) {
-      helper->FreeDirectory(dirp);
+   if (!dirp)
+      return;
+   auto tsfd = static_cast<FindFileData_t *>(dirp);
+   if (TSystem *helper = FindHelper(0, tsfd->fSearchFile)) {
+      helper->FreeDirectory(tsfd->fSearchFile);
       return;
    }
-
-   if (dirp) {
-      ::FindClose(dirp);
-   }
+   ::FindClose(tsfd->fSearchFile);
+   delete tsfd;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 /// Returns the next directory entry.
 
 const char *TWinNTSystem::GetDirEntry(void *dirp)
 {
-   TSystem *helper = FindHelper(0, dirp);
-   if (helper) {
-      return helper->GetDirEntry(dirp);
+   if (!dirp)
+      return nullptr;
+   auto tsfd = static_cast<FindFileData_t *>(dirp);
+   if (TSystem *helper = FindHelper(0, tsfd->fSearchFile)) {
+      return helper->GetDirEntry(tsfd->fSearchFile);
```

Comment:
It seems that of the is a helper you should be passing `dirp` to GetDirEntry? It needs to be symmetrical to OpenDirectory...

Reply:
Then it should be
```
if (TSystem *helper = FindHelper(0, dirp)) {
      return helper->GetDirEntry(dirp);
```
everywhere I suppose?

Reply:
should be fixed now